src/treelearner/monotone_constraints.hpp

```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
+      if (!update_max_constraints) {
+        something_changed = entries_[leaf_idx].UpdateMinAndReturnBoolIfChanged(
+            min_max_constraints.second);
+      } else {
+        something_changed = entries_[leaf_idx].UpdateMaxAndReturnBoolIfChanged(
+            min_max_constraints.first);
+      }
+      // If constraints were not updated, then there is no need to update the leaf
+      if (!something_changed) {
+        return;
+      }
+      leaves_to_update_.push_back(leaf_idx);
+
+    } else { // if node
+      // check if the children are contiguous with the original leaf
+      std::pair<bool, bool> keep_going_left_right = ShouldKeepGoingLeftRight(
+          tree, node_idx, features_of_splits_going_up_from_original_leaf,
+          thresholds_of_splits_going_up_from_original_leaf,
+          was_original_leaf_right_child_of_split);
+      int inner_feature = tree->split_feature_inner(node_idx);
+      uint32_t threshold = tree->threshold_in_bin(node_idx);
+      bool is_split_numerical = tree->IsNumericalSplit(node_idx);
+      bool use_left_leaf_for_update_right = true;
+      bool use_right_leaf_for_update_left = true;
+      // if the split is on the same feature (categorical variables not supported)
+      // then depending on the threshold,
+      // the current left child may not be contiguous with the original right leaf,
+      // or the current right child may not be contiguous with the original left leaf
+      if (is_split_numerical && inner_feature == split_feature) {
+        if (threshold >= split_threshold) {
+          use_left_leaf_for_update_right = false;
+        }
+        if (threshold <= split_threshold) {
+          use_right_leaf_for_update_left = false;
+        }
+      }
+
+      // go down left
+      if (keep_going_left_right.first) {
+        GoDownToFindLeavesToUpdate(
+            tree, tree->left_child(node_idx),
+            features_of_splits_going_up_from_original_leaf,
+            thresholds_of_splits_going_up_from_original_leaf,
+            was_original_leaf_right_child_of_split, update_max_constraints,
+            split_feature, split_info, use_left_leaf,
+            use_right_leaf_for_update_left && use_right_leaf, split_threshold,
+            best_split_per_leaf);
+      }
+      // go down right
+      if (keep_going_left_right.second) {
+        GoDownToFindLeavesToUpdate(
+            tree, tree->right_child(node_idx),
+            features_of_splits_going_up_from_original_leaf,
+            thresholds_of_splits_going_up_from_original_leaf,
+            was_original_leaf_right_child_of_split, update_max_constraints,
+            split_feature, split_info,
+            use_left_leaf_for_update_right && use_left_leaf, use_right_leaf,
+            split_threshold, best_split_per_leaf);
+      }
+    }
+  }
+
+  void
+  GoUpToFindLeavesToUpdate(const Tree* tree, int node_idx, int split_feature,
```

Comment:
the difference between two `GoUpToFindLeavesToUpdate`?

Reply:
if there are similar, could them be merged into one function? if not, maybe different names for them are better.

Reply:
The functions are doing the same thing, I was just using function overloading to make the call simpler. But I can merge the functions no problem.

Reply:
I moved that to the Update function, let me know if you are happy with it. Thanks